### PR TITLE
Gradient accumulation with the distributed trainer

### PR DESCRIPTION
**Pull request by dirkgr**
I re-did gradient accumulation with the distributed trainer.

### CODE REVIEW

File: allennlp/training/trainer.py [hunk 1]
```diff
@@ -343,21 +353,20 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
             train_generator_tqdm = train_generator
 
         cumulative_batch_size = 0
-        for batch in train_generator_tqdm:
+        for batches in lazy_groups_of(train_generator_tqdm, self._num_gradient_accumulation_steps):
```
Comment (brendan-ai2) [inline allennlp/training/trainer.py]: Nit: Maybe `accumulation_groups` (or `batch_group` or whatever) instead of `batches`? Might be confusing to a reader see lines like `cur_batch = sum(training_util.get_batch_size(batch) for batch in batches)`.

File: allennlp/training/trainer.py [hunk 2]
```diff
@@ -323,7 +331,9 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
 
         # Get tqdm for the training batches
         train_generator = self.iterator(self.train_data, num_epochs=1, shuffle=self.shuffle)
```
Comment (brendan-ai2) [inline allennlp/training/trainer.py]: The `lazy_groups_of` call should be here to create a new generator and `tqdm` should be applied to that. As in https://github.com/allenai/allennlp/blob/3dda5ac9fcf390d8b83eb855249360711707c11c/allennlp/training/trainer.py#L306 previously. Otherwise tqdm will give an inaccurate count of the batches.

**Comment by dirkgr** [comment]
@brendan-ai2, I took some license with the naming, and fixed the `tqdm` issue while I was at it.
